nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-controller-service/src/main/java/org/apache/nifi/controller/kudu/KuduLookupService.java

```diff
@@ -0,0 +1,323 @@
+
+        //Scanner
+        KuduScanner.KuduScannerBuilder builder = kuduClient.newScannerBuilder(table);
+
+        builder.setProjectedColumnNames(columnNames);
+
+        //Only expecting one match
+        builder.limit(1);
+
+        coordinates.forEach((key,value)->
+                builder.addPredicate(KuduPredicate.newComparisonPredicate(tableSchema.getColumn(key), KuduPredicate.ComparisonOp.EQUAL, value))
+        );
+
+        KuduScanner kuduScanner = builder.build();
+
+        //Run lookup
+        for ( RowResult row : kuduScanner){
+            final Map<String, Object> values = new HashMap<>();
+            for(String columnName : columnNames){
+                Object object = row.getObject(columnName);
+                if(row.getColumnType(columnName) == Type.BINARY){
+                    object = Base64.getEncoder().encodeToString(row.getBinaryCopy(columnName));
+                }
+                values.put(columnName, object);
+            }
+            return Optional.of(new MapRecord(resultSchema, values));
+        }
+
+        //No match
+        return Optional.empty();
+    }
+    private List<String> getColumns(String columns){
+        if(columns.equals("*")){
+            return tableSchema
+                    .getColumns()
+                    .stream().map(ColumnSchema::getName)
+                    .collect(Collectors.toList());
+        } else {
+            return Arrays.asList(columns.split(","));
+        }
+    }
+    private RecordSchema kuduSchemaToNiFiSchema(Schema kuduTableSchema, List<String> columnNames){
+        final List<RecordField> fields = new ArrayList<>();
+        for(String columnName : columnNames) {
+            if(!kuduTableSchema.hasColumn(columnName)){
+                throw new IllegalArgumentException("Column not found in Kudu table schema " + columnName);
+            }
+            ColumnSchema cs = kuduTableSchema.getColumn(columnName);
+            switch (cs.getType()) {
+                case INT8:
+                    fields.add(new RecordField(cs.getName(), RecordFieldType.BYTE.getDataType()));
+                    break;
+                case INT16:
+                    fields.add(new RecordField(cs.getName(), RecordFieldType.SHORT.getDataType()));
+                    break;
+                case INT32:
+                    fields.add(new RecordField(cs.getName(), RecordFieldType.INT.getDataType()));
+                    break;
+                case INT64:
+                    fields.add(new RecordField(cs.getName(), RecordFieldType.LONG.getDataType()));
+                    break;
+                case UNIXTIME_MICROS:
+                    fields.add(new RecordField(cs.getName(), RecordFieldType.TIMESTAMP.getDataType()));
+                    break;
+                case BINARY:
+                case STRING:
+                    fields.add(new RecordField(cs.getName(), RecordFieldType.STRING.getDataType()));
+                    break;
+                case DECIMAL:
```

Comment:
I don't think it's safe to treat this as a double given the loss of precision that could occur. Converting to string, the same way you handle binary data, looks like the safest approach until NiFi has a Decimal RecordFieldType.